modules/l4/l4_redirect_loopback.c

```diff
@@ -25,89 +26,66 @@ static uint16_t redirect_loopback_process(
 	void **objs,
 	uint16_t nb_objs
 ) {
-	struct rte_ipv4_hdr *ip;
 	struct rte_mbuf *mbuf;
 	struct mbuf_data *d;
-	int edge = NO_IFACE;
+	rte_edge_t edge;
 
 	for (uint16_t i = 0; i < nb_objs; i++) {
 		mbuf = objs[i];
+		edge = REDIRECT;
+
 		d = mbuf_data(mbuf);
 		d->iface = get_vrf_iface(d->iface->vrf_id);
-		if (d->iface)
-			edge = REDIRECT;
-		rte_pktmbuf_prepend(mbuf, sizeof(*ip));
+		if (!d->iface) {
+			edge = NO_IFACE;
+			goto next;
+		}
+
+		if (mbuf->packet_type & RTE_PTYPE_L3_IPV4) {
+			struct ip_local_mbuf_data *d = ip_local_mbuf_data(mbuf);
+			struct rte_ipv4_hdr *ip;
+			ip = (struct rte_ipv4_hdr *)rte_pktmbuf_prepend(mbuf, sizeof(*ip));
+			ip->src_addr = d->src;
+			ip->dst_addr = d->dst;
+			ip->total_length = rte_cpu_to_be_16(d->len) + sizeof(*ip);
+			ip->next_proto_id = d->proto;
+		} else if (mbuf->packet_type & RTE_PTYPE_L3_IPV6) {
+			struct ip6_local_mbuf_data *d = ip6_local_mbuf_data(mbuf);
+			struct rte_ipv6_hdr *ip;
+			ip = (struct rte_ipv6_hdr *)rte_pktmbuf_prepend(mbuf, sizeof(*ip));
+			ip->src_addr = d->src;
+			ip->dst_addr = d->dst;
+			ip->payload_len = rte_cpu_to_be_16(d->len);
+			ip->hop_limits = d->hop_limit;
+			ip->proto = d->proto;
+		} else {
+			edge = BAD_PROTO;
+		}
+
+next:
 		if (gr_mbuf_is_traced(mbuf)) {
 			gr_mbuf_trace_add(mbuf, node, 0);
 		}
+		rte_node_enqueue_x1(graph, node, edge, mbuf);
 	}
 
-	rte_node_enqueue(graph, node, edge, objs, nb_objs);
 	return nb_objs;
 }
 
-static struct rte_node_register tcp_redirect_loopback_node = {
-	.name = "tcp_redirect_loopback",
-	.process = redirect_loopback_process,
-	.nb_edges = EDGE_COUNT,
-	.next_nodes = {
-		[REDIRECT] = "loopback_output",
-		[NO_IFACE] = "no_loop_iface",
-	},
-};
-
-static struct rte_node_register udp_redirect_loopback_node = {
-	.name = "udp_redirect_loopback",
-	.process = redirect_loopback_process,
-	.nb_edges = EDGE_COUNT,
-	.next_nodes = {
-		[REDIRECT] = "loopback_output",
-		[NO_IFACE] = "no_loop_iface",
-	},
-};
-
```

Comment:
I liked the idea of having two separate nodes for tcp and udp. We may not want to register the same local ports depending on the protocol.

Maybe these ports should be renamed to tcp_input_local and udp_input_local which would (or would not) redirect to loopback_output after reconstructing the ip(v6) headers.

Reply:
This is already the case, the register function already takes a proto as a parameter, so we can select TCP/UDP/...
We only support udp port as of now, as I'm not yet aware of a TCP based tunnel.
All non-handled traffic will be redirected to the loop interface.
Even this behavior may have to be refined, in a future patch.